src/rgw/rgw_pubsub.h

```diff
@@ -399,20 +401,23 @@ struct rgw_pubsub_topic {
   std::string arn;
   std::string opaque_data;
   std::string policy_text;
+  // buckets associated with topic along with their notification id.
+  std::map<rgw_bucket, std::set<std::string> > bucket_notif_id;
```

Comment:
is it really necessary for the topic metadata to track all of the buckets that refer to it?

if that is a requirement, i think it would be preferable if we store that list outside of the topic metadata object, so we don't have to write (and replicate) the topic every time we change a bucket notification policy

as an example, rgw has user metadata, and each user has to track all of the buckets it owns for the ListBuckets op. but we don't store that list of buckets in the user metadata itself. instead, each zone tracks a local list of the user's buckets in a separate {user}.buckets object, and updates that list as buckets are created/deleted/replicated

Reply:
> is it really necessary for the topic metadata to track all of the buckets that refer to it?
> 
> if that is a requirement, i think it would be preferable if we store that list outside of the topic metadata object, so we don't have to write (and replicate) the topic every time we change a bucket notification policy
> 
> as an example, rgw has user metadata, and each user has to track all of the buckets it owns for the ListBuckets op. but we don't store that list of buckets in the user metadata itself. instead, each zone tracks a local list of the user's buckets in a separate {user}.buckets object, and updates that list as buckets are created/deleted/replicated

I see, but then what's the benefit of maintaining a separate rados list object ?you still have to update that list object ? I am guessing just that list object is not replicated for every update, each zone updates maintains that list itself and updates it ? So you saving the replication event ?

Reply:
> > is it really necessary for the topic metadata to track all of the buckets that refer to it?
> > if that is a requirement, i think it would be preferable if we store that list outside of the topic metadata object, so we don't have to write (and replicate) the topic every time we change a bucket notification policy
> > as an example, rgw has user metadata, and each user has to track all of the buckets it owns for the ListBuckets op. but we don't store that list of buckets in the user metadata itself. instead, each zone tracks a local list of the user's buckets in a separate {user}.buckets object, and updates that list as buckets are created/deleted/replicated
> 
> I see, but then what's the benefit of maintaining a separate rados list object ?you still have to update that list object ? I am guessing just that list object is not replicated for every update, each zone updates maintains that list itself and updates it ? So you saving the replication event ?

Also I was thinking, if we maintain a separate list, how do we display that info with topic listing?  while reading topic we  also read this list and populate it?

Reply:
that's right. and the list updates can be done atomically with something like `omap_set()`, whereas updates to this map stored in the topic metadata would require a read-modify-write with retries when we detect a racing write

can you help me understand why the topic needs to track the buckets? this seems like a new requirement

Reply:
> that's right. and the list updates can be done atomically with something like `omap_set()`, whereas updates to this map stored in the topic metadata would require a read-modify-write with retries when we detect a racing write
> 
> can you help me understand why the topic needs to track the buckets? this seems like a new requirement

There are 2 reasons to it
1> when you delete the topic, you need to delete the notifications associated with it, else even with topic deleted bucket notifications will continue try sending the events, as while sending event there is no explicit call to read the topic info, the topic info is stored in bucket notification object itself. 

2> from telemetry perspective, it's nice to know what all buckets are associated  with the topic, so when you list topics you also get the subscribed bucket information

Reply:
topic updates are also going to work much better with this. currently, the updates apply only to new notifications

Reply:
> topic updates are also going to work much better with this. currently, the updates apply only to new notifications

does this imply that topic updates from the rest api will have to loop and potentially modify lots of bucket metadata?

> 1> when you delete the topic, you need to delete the notifications associated with it, else even with topic deleted bucket notifications will continue try sending the events, as while sending event there is no explicit call to read the topic info, the topic info is stored in bucket notification object itself.

maybe it's better to actually read the topic information at the point we generate notifications? topics are cached so the cost is amortized like other types of metadata

Reply:
> > topic updates are also going to work much better with this. currently, the updates apply only to new notifications
> 
> does this imply that topic updates from the rest api will have to loop and potentially modify lots of bucket metadata?
> 
> > 1> when you delete the topic, you need to delete the notifications associated with it, else even with topic deleted bucket notifications will continue try sending the events, as while sending event there is no explicit call to read the topic info, the topic info is stored in bucket notification object itself.
> 
> maybe it's better to actually read the topic information at the point we generate notifications? topics are cached so the cost is amortized like other types of metadata

Not loop, the bucket notification attribute is stored as map with topic name as the key, so it will read only the keys(topic that's updated) and update the topic. But yes it will loop for all the buckets that are subscribed to that topic 

regarding reading the topic for each notification(even with it cached) I feel adding latency to S3 request is little better as it's changing config and latency change should not matter much there for client (latency change due to looping for all subscribed buckets )

Reply:
getting the topic configuration is happening for every op in the fastpath.
while updating and deleting topics is happening way less often. 

currently, we just don't delete/update the notification when the topic changes.

Reply:
> getting the topic configuration is happening for every op in the fastpath.
> while updating and deleting topics is happening way less often.

that's certainly true. but i worry about interdependencies between metadata objects like this, especially if we want to scale to many buckets. if a topic update also has to update many bucket instance objects, what happens if rgw crashes in the middle? and what happens if two separate clients write a topic update, and they race to write different updates to the buckets? we can do atomic updates to one metadata object, but it's much harder to guarantee consistency between related objects

regarding reads in the fast path, we already read user and bucket metadata in each op, and the metadata cache has been very effective there. i don't think it's unreasonable to read topic metadata for ops that have notifications enabled. that makes this consistency issue much simpler because ops always see the latest version of the topic

Reply:
> > getting the topic configuration is happening for every op in the fastpath.
> > while updating and deleting topics is happening way less often.
> 
> that's certainly true. but i worry about interdependencies between metadata objects like this, especially if we want to scale to many buckets. if a topic update also has to update many bucket instance objects, what happens if rgw crashes in the middle? and what happens if two separate clients write a topic update, and they race to write different updates to the buckets? we can do atomic updates to one metadata object, but it's much harder to guarantee consistency between related objects
> 
> regarding reads in the fast path, we already read user and bucket metadata in each op, and the metadata cache has been very effective there. i don't think it's unreasonable to read topic metadata for ops that have notifications enabled. that makes this consistency issue much simpler because ops always see the latest version of the topic

in real world scenario, ideally there would not be many TOPIC updates/deletes. And again multiple user updating the same topic at same time is not real use case with the topic policy in place where users would need policy to access the topic.
but coming back to your concerns.. there are 2 concerns so far...
- keeping mapping of bucket and topic inside each other, so updating/deleting any of them results in updating the other and then replication
- second issue,  bucket notification is not reading the topic  object from rados but using the stored topic info. This model can be problematic when there are rgw crashes while topic is updated or could be problem if there are lots of bucket that subscribed to single topic causing to update multiple buckets for every update. (both concerns are related to topic updates, which in real world is not a common s3 operation).

shall i still go ahead and update the commits with 3 fixes so far discussed (feature renamed, stop using Rados cast and use tenant info in the topic key) ?

Reply:
> > > getting the topic configuration is happening for every op in the fastpath.
> > > while updating and deleting topics is happening way less often.
> > 
> > 
> > that's certainly true. but i worry about interdependencies between metadata objects like this, especially if we want to scale to many buckets. if a topic update also has to update many bucket instance objects, what happens if rgw crashes in the middle? and what happens if two separate clients write a topic update, and they race to write different updates to the buckets? we can do atomic updates to one metadata object, but it's much harder to guarantee consistency between related objects
> > regarding reads in the fast path, we already read user and bucket metadata in each op, and the metadata cache has been very effective there. i don't think it's unreasonable to read topic metadata for ops that have notifications enabled. that makes this consistency issue much simpler because ops always see the latest version of the topic
> 
> in real world scenario, ideally there would not be many TOPIC updates/deletes. And again multiple user updating the same topic at same time is not real use case with the topic policy in place where users would need policy to access the topic. but coming back to your concerns.. there are 2 concerns so far...
> 
> * keeping mapping of bucket and topic inside each other, so updating/deleting any of them results in updating the other and then replication
> * second issue,  bucket notification is not reading the topic  object from rados but using the stored topic info. This model can be problematic when there are rgw crashes while topic is updated or could be problem if there are lots of bucket that subscribed to single topic causing to update multiple buckets for every update. (both concerns are related to topic updates, which in real world is not a common s3 operation).
> 
> shall i still go ahead and update the commits with 3 fixes so far discussed (feature renamed, stop using Rados cast and use tenant info in the topic key) ?

i have updated the commits with those 3 changes

Reply:
thanks for clarifying, i tend to agree with @cbodley now...
even tough it is unlikely, the lack of atomicity, may leave the system in an invalid state, that would be very difficult to debug or recover from.
for getting the list of buckets tied to a specific topic, we can use a separate object (per topic) that will be used only  from `radosgw-admin`. we can limit that to the master zone, so syncing may not be needed for that object.

Reply:
> thanks for clarifying, i tend to agree with @cbodley now... even tough it is unlikely, the lack of atomicity, may leave the system in an invalid state, that would be very difficult to debug or recover from. (this will also simplify the implementation) for getting the list of buckets tied to a specific topic, we can use a separate object (per topic) that will be used only from `radosgw-admin`. we can limit that to the master zone, so syncing may not be needed for that object.

So thinking more on this, currently the `bucket_notif` member that i have added in current PR to track the buckets for a given topic gets updated only during the `Creation/Deletion of Bucket notification`.  if we store the mapping in new rados object, how will this work for secondary zone ? during the the create/delete bucket notification event the secondary zone will just update the attrs that is triggered from metadata sync, so we now have to read the attrs every time in secondary zone and check if there is any update to `BUCKET_NOTIFICATION` and if yes then go and update the new topic object in secondary zone?
is there a way for `merge_store_attrs()` to indicate that this attr update was for the `BUCKET_NOTIFICATION` attribute, if yes it would be easy to setup a separate metadata topic object but if not then this could be a  non-trivial?

Reply:
if we don't use the bucket list to do updates and deletion, and we lookup the topic from the notification (as @cbodley suggested).
it means that the bucket list is only used for debugging from `radosgw-admin` command.
maybe this system object is local to each zone and the admin command is used to gather all info?

Reply:
> it means that the bucket list is only used for debugging

I am not sure if thats what casey was suggesting, i interpret it, that he was suggesting to not have a separate member `bucket_notif_id` inside the `rgw_pubsub_topic` struct. That member was storing the bucket topic mappings and instead have a separate rados list object just like how user rados object has a separate `user-bucket-list` object which is use to look up buckets for a given `user`. And that separate rados object is not synced across the zones, but each zone maintains its own list object. (And i had concern about how will secondary zone build this list object as this list object would be created/updated/deleted during the create/delete bucket notification operations and during those operation the secondary zone does not know if the ATTR update is for `BUCKET_NOTIFICATION`.
Coming back to your understanding, you are suggesting that topics rados object should not be used at all and we rather user bucket_notification attribute to list all topics ?

Reply:
this buckets-per-topic index is a new requirement added by this pr. the s3/sns APIs themselves don't seem to require it, so i'd argue that it's not worth the complexity to maintain. without this index, it's still possible (though expensive at scale) to enumerate all bucket instances to see which refer to a given topic

above i pointed out an example where a similar metadata index is required by s3. for each user, we create a separate {userid}.buckets object that tracks the buckets owned by that user to serve its ListBuckets requests. we also use that object to accumulate bucket stats for user quota enforcement and reporting via 'radosgw-admin user stats'

this {userid}.buckets object gets created on each zone in multisite, but isn't directly replicated by metadata sync. it's just updated on any changes to the bucket metadata, whether written locally or by [RGWBucketMetadataHandler](https://github.com/ceph/ceph/blob/c52c2dca6/src/rgw/driver/rados/rgw_bucket.cc#L2108-L2118) upon replication

Reply:
> this buckets-per-topic index is a new requirement added by this pr. the s3/sns APIs themselves don't seem to require it, so i'd argue that it's not worth the complexity to maintain. without this index, it's still possible (though expensive at scale) to enumerate all bucket instances to see which refer to a given topic
> 
> above i pointed out an example where a similar metadata index is required by s3. for each user, we create a separate {userid}.buckets object that tracks the buckets owned by that user to serve its ListBuckets requests. we also use that object to accumulate bucket stats for user quota enforcement and reporting via 'radosgw-admin user stats'
> 
> this {userid}.buckets object gets created on each zone in multisite, but isn't directly replicated by metadata sync. it's just updated on any changes to the bucket metadata, whether written locally or by [RGWBucketMetadataHandler](https://github.com/ceph/ceph/blob/c52c2dca6/src/rgw/driver/rados/rgw_bucket.cc#L2108-L2118) upon replication

@cbodley, TBH there isn't that much of a complexity, you update `bucket_notif_id` only during the notification create/delete s3 operations and that will trigger a topic update asking the secondary to sync the topic along with bucket notifications( atomicity is something you pointed could be problem). Maintaining a separate rados list object like `user-bucket` does not buy much, coz then to look up buckets mapping for topic you still end up reading the new list object and also how will the secondary zone create/update this new list object as this ATTR update does not tell us today which ATTR is getting updated, so secondary zone has to do a match between previous and current `BUCKET_NOTIFICATION` attr  for every attr update and if thats modified then go an update/create the list object.

Reply:
sharing a link to multisite's handling of [RGW_ATTR_LC changes](https://github.com/ceph/ceph/blob/c52c2dca6/src/rgw/driver/rados/rgw_bucket.cc#L2611-L2638) to update the local lc list

Reply:
> sharing a link to multisite's handling of [RGW_ATTR_LC changes](https://github.com/ceph/ceph/blob/c52c2dca6/src/rgw/driver/rados/rgw_bucket.cc#L2611-L2638) to update the local lc list

Thanks for sharing @cbodley , i was just checking this code and from the first look i see it also use cls_interface to store the list and second there is no comparison between current and old attr, as long as there is `LC_ATTR`, it goes and updates the list object.